python/lsst/afw/geom/endpoint.cc

```diff
@@ -0,0 +1,182 @@
+/*
+* LSST Data Management System
+* See COPYRIGHT file at the top of the source tree.
+*
+* This product includes software developed by the
+* LSST Project (http://www.lsst.org/).
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the LSST License Statement and
+* the GNU General Public License along with this program. If not,
+* see <http://www.lsstcorp.org/LegalNotices/>.
+*/
+#include <ostream>
+#include <memory>
+
+#include "pybind11/pybind11.h"
+#include "pybind11/stl.h"
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/SpherePoint.h"
+#include "lsst/afw/geom/Endpoint.h"
+
+namespace py = pybind11;
+using namespace py::literals;
+
+namespace lsst {
+namespace afw {
+namespace geom {
+namespace {
+
+template <typename PyClass>
+void addStrAndRepr(PyClass& cls) {
+    using Class = typename PyClass::type;  // C++ class associated with pybind11 wrapper class
+    cls.def("__repr__", [](Class const& self) {
+        std::ostringstream os;
+        os << "lsst.afw.geom." << self;
+        return os.str();
+    });
+    cls.def("__str__", [](Class const& self) {
+        std::ostringstream os;
+        os << self;
+        return os.str();
+    });
+}
+
+// Declare BaseVectorEndpoint<Point, Array>;
+// this is meant to be called by other `declare...` functions;
+template <typename Point, typename Array>
+void declareBaseEndpoint(py::module& mod, std::string const& suffix) {
+    using Class = BaseEndpoint<Point, Array>;
+
+    py::class_<Class, std::shared_ptr<Class>> cls(mod, ("_BaseEndpoint" + suffix).c_str());
+
+    // abstract base class so do not wrap the constructor
+
+    cls.def("getNAxes", &Class::getNAxes);
+    cls.def("normalizeFrame", &Class::normalizeFrame);
+    // return a deep copy so Python cannot modify the internal state
+    cls.def("makeFrame", [](Class const & self) {
+        auto frame = self.makeFrame();
+        return frame->copy();
+    });
```

Comment:
I didn't wrap `getNPoints` and etc. because they were not implemented in this base class, but I believe it's OK to wrap them anyway so I'll try that.

I don't anticipate us making Python-only subclasses, and the protected methods aren't essential to doing so.